File: lsif-java/src/main/scala/com/sourcegraph/lsif_java/buildtools/TemporaryFiles.scala
package com.sourcegraph.lsif_java.buildtools

import java.nio.file.Files
import java.nio.file.Path

import com.sourcegraph.io.DeleteVisitor
import com.sourcegraph.lsif_java.commands.IndexCommand

object TemporaryFiles {
  def withDirectory[T](index: IndexCommand)(fn: Path => T): T = {
    index.temporaryDirectory match {
      case Some(tmp) =>
        fn(tmp)
      case None =>
        val tmp = Files.createTempDirectory("lsif-java")
        try fn(tmp)
        finally {
          if (index.cleanup) {
            Files.walkFileTree(tmp, new DeleteVisitor())
          }
        }
    }
  }
}

File: scip-java/src/main/kotlin/com/sourcegraph/scip_java/buildtools/TemporaryFiles.kt
package com.sourcegraph.scip_java.buildtools

import com.sourcegraph.scip_java.commands.IndexCommand
import java.nio.file.Files
import java.nio.file.Path

object TemporaryFiles {
    fun <T> withDirectory(index: IndexCommand, fn: (Path) -> T): T {
        val explicit = index.temporaryDirectory
        if (explicit != null) {
            return fn(explicit)
        }
        val tmp = Files.createTempDirectory("scip-java")
        try {
            return fn(tmp)
        } finally {
            if (index.cleanup) {
                tmp.toFile().deleteRecursively()
            }
        }
    }
}
